Alerting: update authorization logic to use proper legacy roles when fine-grained access is disabled

yuri-tceretian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
When fine-grained access is disabled, the authorization logic falls back to legacy authorization, which is by default requires user to be signed in. The additional logic was implemented on request handler side (not every where). 

This PR updates the authorization method to require `Editor` role for all POST, PUT, DELETE requests by default. And Viewer role for all other request types. There are some overrides for particular paths. Some of them were added to match the legacy behavior. 

I had to return handler ReqSignedIn to the routes group because web handler `middleware.ReqEditorRole` does not check that user is not authenticated, and therefore they get wrong status code in the response.

yuri-tceretian committed:
require legacy Editor for post, put, delete endpoints

yuri-tceretian committed:
remove role check in request handlers

gotjosh commented on pkg/services/ngalert/api/authorization.go:
```diff
@@ -136,13 +150,15 @@ func (api *API) authorize(method, path string) web.Handler {
 		// additional authorization is done in the request handler
 		eval = ac.EvalAny(ac.EvalPermission(ac.ActionAlertingNotificationsUpdate), ac.EvalPermission(ac.ActionAlertingNotificationsCreate), ac.EvalPermission(ac.ActionAlertingNotificationsDelete))
 	case http.MethodPost + "/api/alertmanager/grafana/config/api/v1/receivers/test":
+		fallback = ac.ReqOrgAdminOrEditor // moved from the request handler as is
```
```suggestion
		fallback = ac.ReqOrgAdminOrEditor
```

gotjosh replied:
Not sure these comments make sense, what are you trying to convey? Will they be useful in the long run?

gotjosh commented on pkg/services/ngalert/api/authorization.go:
```diff
@@ -26,6 +26,17 @@ func (api *API) authorize(method, path string) web.Handler {
 	authorize := acmiddleware.Middleware(api.AccessControl)
 	var eval ac.Evaluator = nil
 
+	// can be overridden in path matchers below
```
```suggestion
	// Most routes follow this general authorization approach as a fallback. Exceptions are overridden directly in the below block.

gotjosh commented on pkg/services/ngalert/api/authorization.go:
```diff
@@ -136,13 +150,15 @@ func (api *API) authorize(method, path string) web.Handler {
 	case http.MethodDelete + "/api/alertmanager/grafana/config/api/v1/alerts": // reset alertmanager config to the default
 		eval = ac.EvalPermission(ac.ActionAlertingNotificationsDelete)
 	case http.MethodGet + "/api/alertmanager/grafana/config/api/v1/alerts":
+		fallback = ac.ReqOrgAdminOrEditor // moved from the request handler as is
 		eval = ac.EvalPermission(ac.ActionAlertingNotificationsRead)
```
This is somewhat confusing.

My understanding is that `eval` would only contain the permission if FGAC is enabled. However, you're assigning two variables when you know ahead of time that only of them might not return something. 

While somewhat innocuous is not very obvious and can lead to harder debugging. How do you feel about something like:

```golang
type scopedAction struct {
	action string
	scopes []string
}

func evalWithFallback(actions []scopedAction, fallback string) Evaluator {
	if len(actions) == 1 {
		// evaluate one
	}

	// evaluateAny

	return fallback
}
```

With that, you could replace this to:

```golang
evalWithFallback(ac.ActionAlertingNotificationsRead, ac.ReqOrgAdminOrEditor)
```

yuri-tceretian replied:
exactly this happens several line below 
https://github.com/grafana/grafana/pull/46931/files#diff-627a793303894bbf1b53347d4c9c0316e28159fa392b13786f80553dc71e8dd4R185

This method called only when Grafana server starts and registers API routes. Once per route.

gotjosh replied:
Discussed offline, I misunderstood what was going on here.

gotjosh reviewed (commented):
There's probably bigger fish to fry in `authorization.go` but please see my comment as I think it can be a quick win for routes that have a fallback.

yuri-tceretian committed:
fix web handlers

yuri-tceretian committed:
require user to be signed in on group level

because handler that checks that user has role Editor does not check it is signed in

yuri-tceretian committed:
fix intergation tests. just messages

yuri-tceretian committed:
fix comment

yuri-tceretian committed:
Merge branch 'up/main' into fgac-legacy-role-to-authz
